serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/MatrixClient.kt

```diff
@@ -313,6 +314,86 @@ class MatrixSession(val client: HttpClient, val server: String, val user: String
+        val (displayname, avatar_url) =
+                if(roomId != null) { //Get info at the Room-Member level
+                     when (val room_member_info = getLocalRoomMemberDetails(sender, roomId)) {
+                        is Success -> Pair(room_member_info.value.first, room_member_info.value.second)
+                        is Error -> Pair(sender, null)
+                    }
+                } else { //Get info at the User-Profile level
+                    when (val user_profile_info = getLocalUserProfileDetails(sender)) {
+                        is Success -> Pair(user_profile_info.value.first, user_profile_info.value.second)
+                        is Error -> Pair(sender, null)
+                    }
+                }
+
+        return if (avatar_url != null) {
+            when (val avatar_file_path = getLocalMediaPathFromUrl(avatar_url)) {
+                is Success -> Pair(displayname, avatar_file_path.value)
+                is Error -> Pair(displayname, null)
+            }
+        } else {
+            Pair(displayname, null)
+        }
+    }
+
+    fun getLocalUserProfileDetails(sender: String): Outcome<Pair<String?, String?>> {
+        try {
+            val cached_user = Database.getUserProfileFromCache(sender)
+            if (cached_user != null) {
+                return Success(cached_user)
+            }
+
+            synchronized(this) {
+                if (in_flight_user_requests.contains(sender)) {
+                    return Error("User request already in flight");
+                }
+                in_flight_user_requests.add(sender)
+            }
+            thread(start = true) {
+                //No valid cache hit
+                println("Background thread request for $sender")
+                val (displayname, avatar_url) = getUserProfile(sender)
+                val result = Database.addUserProfileToCache(sender, displayname, avatar_url, false)
+                println("Finished downloading $sender in the background with $result")
+                synchronized(this) {
+                    in_flight_user_requests.remove(sender)
+                }
+                onUpdate()
+            }
+
+            println("Returning early, downloading $sender in the background")
+            return Error("Downloading user profile in the background")
+        } catch (e: Exception) {
+            println("Error with user profile retrieval $e")
+            return Error("User Profile Retrieval Failed", e)
+        }
+    }
+
+    fun getLocalRoomMemberDetails(sender: String, roomId: String): Outcome<Pair<String?, String?>> {
+        return try {
+            val roomMemberEventContent = Database.getStateEvent(session_id, roomId, "m.room.member", sender)
+                    ?.castToStateEventWithContentOfType<RoomMemberEventContent>()
+            if(roomMemberEventContent == null) {
+                Success(Pair(sender, null))
```

Comment:
I might make this Error instead of success, since we didn't properly pull an actual name. The calling function will still fall back to sender ID, so it doesn't make a functional difference at this time though, so ehh?